feat: 🎸 Add basic setup to support confidential assets

prashantasdeveloper opened the pull request:
### Description

* Updates the polkadot types with confidential asset pallets and various related events and extrinsics
* Adds ConfidentialAsset entity. For now the entity has basic `ID` and couple of methods `details` and `exists` to get the basic info about CA
* Adds ConfidentialAssets namespace which for now has a single method that allows getting any CA with its ID

### Breaking Changes

NA

### JIRA Link

DA-984
DA-985
DA-986
DA-1006

### Checklist

- [ ] Updated the Readme.md (if required) ?

polymath-eric commented on src/api/entities/confidential/ConfidentialAsset/types.ts:
```diff
@@ -0,0 +1,10 @@
+import { Identity } from '~/internal';
+
+export interface ConfidentialAssetDetails {
+  ticker?: string;
```
why is ticker not always defined?

I think it would be a good place to comment why if its intended

polymath-eric commented on src/utils/internal.ts:
```diff
@@ -1891,3 +1892,28 @@ export function areSameClaims(
+  if (id.length >= 32) {
+    let guid = id;
+
+    if (!Guid.isGuid(id)) {
```
```suggestion
    const uuidRegex = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/
    
     if (!uuidRegex.test(id)) {
```

polymath-eric replied:
There are multiple open issues regarding this function btw: https://github.com/snico-dev/guid-typescript/issues

polymath-eric commented on src/api/entities/confidential/ConfidentialAsset/index.ts:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * GUID of the asset
+   *
+   * @note the value can either be a valid guid like `76702175-d8cb-e3a5-5a19-734433351e26` or can be a string representing the guid without the `-` like `76702175d8cbe3a55a19734433351e26`
```
After reading through the PR, I am thinking these should just be plain hex values.

These should be chain generated, and I am assuming all 128 bits are random. We shouldn't insert the `-` and call them GUIDs or UUIDs unless they conform to the actual specs. Both of those terms refer to something well defined, and its misleading to label random bits and them.

https://en.wikipedia.org/wiki/Universally_unique_identifier#Versions

polymath-eric commented on src/utils/internal.ts:
```diff
@@ -1891,3 +1892,28 @@ export function areSameClaims(
+/**
+ * @hidden
+ */
+export function assertCaAssetValid(id: string): string {
```
this isn't a simple assert, but it transforms the value, so shouldn't have the `assert` prefix.

Unless I am  missing something we should not insert the `-`. An actual UUID doesn't seem good for privacy. UUIDs generated on one computer can get grouped together with v4, v7 has time sorting, etc.

polymath-eric reviewed (changes requested):
LGMT, but many pedantic comments around GUID/UUID terminology.

If the IDs are truly random 128 bits, they are not technically UUIDs. If they are valid UUIDs, then there are privacy concerns since they aren't fully random.

polymath-eric commented on package.json:
```diff
@@ -83,6 +83,7 @@
     "eslint-plugin-promise": "^6.1.1",
     "eslint-plugin-simple-import-sort": "^10.0.0",
     "fs-extra": "11.1.1",
+    "guid-typescript": "^1.0.9",
```
not `crypto.randomUUID()`, will give a unique identifier, but not every bit in it will be truly random.

prashantasdeveloper replied:
I have removed the dep

prashantasdeveloper commented on package.json:
```diff
@@ -83,6 +83,7 @@
     "eslint-plugin-promise": "^6.1.1",
     "eslint-plugin-simple-import-sort": "^10.0.0",
     "fs-extra": "11.1.1",
+    "guid-typescript": "^1.0.9",
```
I have removed the dep

prashantasdeveloper commented on src/api/entities/confidential/ConfidentialAsset/index.ts:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * GUID of the asset
+   *
+   * @note the value can either be a valid guid like `76702175-d8cb-e3a5-5a19-734433351e26` or can be a string representing the guid without the `-` like `76702175d8cbe3a55a19734433351e26`
```
IMO we don't need to go into details of uuid here. Consider it as a string value, but to be in specific format. It's easier for end user to understand that guid has specific format and it has to add in that way. 

I could have kept just the string without the hyphen but Robert's proof server requires the asset ID to be passed in this way even when we have to pass in plain format to the chain (with 0x prefixed)

prashantasdeveloper commented:
@polymath-eric I kept the uuid format supportive because that's what Robert's proof server spits out. And someone testing things out might make use of those Asset IDs. 

Also, I have removed references of guid from docs and removed the added dependency as well. 

We can circle back on this whether to continue this or just support it is plain string

prashantasdeveloper commented on src/polkadot/augment-api-events.ts:
```diff
@@ -2284,6 +2442,20 @@ declare module '@polkadot/api-base/types/events' {
+       * The \[sudoer\] just switched identity; the old key is supplied.
+       **/
+      KeyChanged: AugmentedEvent<ApiType, [Option<AccountId32>]>;
+      /**
+       * A sudo just took place. \[result\]
+       **/
+      Sudid: AugmentedEvent<ApiType, [Result<Null, SpRuntimeDispatchError>]>;
+      /**
+       * A sudo just took place. \[result\]
+       **/
+      SudoAsDone: AugmentedEvent<ApiType, [Result<Null, SpRuntimeDispatchError>]>;
```
The types will most probably again be updated. So leaving as is

prashantasdeveloper commented:
:tada: This PR is included in version 24.0.0-confidential-assets.1 :tada:

The release is available on:
- [npm package (@confidential-assets dist-tag)](https://www.npmjs.com/package/@polymeshassociation/polymesh-sdk/v/24.0.0-confidential-assets.1)
- [GitHub release](https://github.com/PolymeshAssociation/polymesh-sdk/releases/tag/v24.0.0-confidential-assets.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
